File: razer-sdk/src/test/java/net/booone/razersdk/effects/ProgressiveRzKeySelectorEffectTest.java
package net.booone.razersdk.effects;

import net.booone.razersdk.ChromaRestSDK;
import net.booone.razersdk.animation.SimpleFrame;
import net.booone.razersdk.color.StaticColor;
import net.booone.razersdk.sdk.ProgressiveRzKeySelector;
import net.booone.razersdk.sdk.RzKey;
import net.booone.razersdk.sdk.RzKeySelector;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;
import java.util.stream.IntStream;

public class ProgressiveRzKeySelectorEffectTest {
    @Test
    void testProgressiveRzKeySelectorAnimation() throws InterruptedException {
        List<Set<RzKey>> parts = IntStream.range(RzKey.RZKEY_TILDE.getColumn(), RzKey.RZKEY_BACKSLASH.getColumn() + 1)
                .mapToObj(col -> new RzKeySelector().withColumn(col).withRowBetween(RzKey.RZKEY_TILDE, RzKey.RZKEY_LCTRL).asSet())
                .toList();
        try (ChromaRestSDK chromaRestSDK = new ChromaRestSDK()) {
            ProgressiveRzKeySelector progressiveRzKeySelector = new ProgressiveRzKeySelector(parts, 3);

            for (int i = 0; i < 60; i++) {
                chromaRestSDK.createKeyboardEffect(new SimpleFrame(progressiveRzKeySelector.getNextPart(), StaticColor.WHITE));
                Thread.sleep(100);
            }
        }
    }
}

File: test/razer-sdk/effects/ProgressiveRzKeySelectorEffect.test.ts
import {describe, test} from 'vitest';
import {ChromaNativeSDK} from '../../../src/razer-sdk/ChromaNativeSDK.js';
import {SimpleFrame} from '../../../src/razer-sdk/animation/SimpleFrame.js';
import {StaticColor} from '../../../src/razer-sdk/color/StaticColor.js';
import {ProgressiveRzKeySelector} from '../../../src/razer-sdk/sdk/ProgressiveRzKeySelector.js';
import {getColumn, RzKey} from '../../../src/razer-sdk/sdk/RzKey.js';
import {RzKeySelector} from '../../../src/razer-sdk/sdk/RzKeySelector.js';

/**
 * Integration test for progressive column selector effect on actual hardware.
 * Creates a sliding window of columns moving across the keyboard.
 */
describe('ProgressiveRzKeySelectorEffect Integration', () => {
    test('testProgressiveRzKeySelectorAnimation', async () => {
        const sdk = new ChromaNativeSDK();
        await sdk.connect();

        try {
            const startCol = getColumn(RzKey.RZKEY_TILDE);
            const endCol = getColumn(RzKey.RZKEY_BACKSLASH);
            const parts: Set<RzKey>[] = [];
            for (let col = startCol; col <= endCol; col++) {
                parts.push(new RzKeySelector()
                    .withColumn(col)
                    .withRowBetween(RzKey.RZKEY_TILDE, RzKey.RZKEY_LCTRL)
                    .asSet());
            }
            const progressiveRzKeySelector = new ProgressiveRzKeySelector(parts, 3);

            for (let i = 0; i < 60; i++) {
                const nextPart = progressiveRzKeySelector.getNextPart();
                await sdk.createKeyboardEffect(new SimpleFrame([...nextPart], StaticColor.WHITE));
                await new Promise(resolve => setTimeout(resolve, 100));
            }
        } finally {
            await sdk.close();
        }
    }, 30000);
});
